Flaky-test: ExtensibleLoadManagerCloseTest.testCloseAfterLoadingBundles

### Search before reporting

- [x] I searched in the [issues](https://github.com/apache/pulsar/issues) and found nothing similar.


### Example failure

https://github.com/apache/pulsar/actions/runs/16959261008/job/48081019260?pr=24626#step:11:1308

### Exception stacktrace

```text
Error:  Tests run: 13, Failures: 1, Errors: 0, Skipped: 4, Time elapsed: 393.24 s <<< FAILURE! - in org.apache.pulsar.broker.loadbalance.extensions.ExtensibleLoadManagerCloseTest
  Error:  org.apache.pulsar.broker.loadbalance.extensions.ExtensibleLoadManagerCloseTest.testCloseAfterLoadingBundles  Time elapsed: 3.172 s  <<< FAILURE!
  org.apache.pulsar.client.admin.PulsarAdminException$ServerSideErrorException: 
  
   --- An unexpected error occurred in the server ---
  
  Message: Failed to tableview on non-persistent://pulsar/system/loadbalancer-broker-load-data, restart reason: object is null
  
  Stacktrace:
  
  java.lang.IllegalStateException: Failed to tableview on non-persistent://pulsar/system/loadbalancer-broker-load-data, restart reason: object is null
  	at org.apache.pulsar.broker.loadbalance.extensions.store.TableViewLoadDataStoreImpl.get(TableViewLoadDataStoreImpl.java:98)
  	at org.apache.pulsar.broker.loadbalance.extensions.strategy.LeastResourceUsageWithWeight.select(LeastResourceUsageWithWeight.java:101)
  	at org.apache.pulsar.broker.loadbalance.extensions.ExtensibleLoadManagerImpl.lambda$selectAsync$19(ExtensibleLoadManagerImpl.java:643)
  	at java.base/java.util.concurrent.CompletableFuture.uniApplyNow(CompletableFuture.java:684)
  	at java.base/java.util.concurrent.CompletableFuture.uniApplyStage(CompletableFuture.java:662)
  	at java.base/java.util.concurrent.CompletableFuture.thenApply(CompletableFuture.java:2200)
  	at org.apache.pulsar.broker.loadbalance.extensions.ExtensibleLoadManagerImpl.lambda$selectAsync$20(ExtensibleLoadManagerImpl.java:638)
  	at java.base/java.util.concurrent.CompletableFuture$UniCompose.tryFire(CompletableFuture.java:1150)
  	at java.base/java.util.concurrent.CompletableFuture$Completion.exec(CompletableFuture.java:483)
  	at java.base/java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:387)
  	at java.base/java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(ForkJoinPool.java:1312)
  	at java.base/java.util.concurrent.ForkJoinPool.scan(ForkJoinPool.java:1843)
  	at java.base/java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1808)
  	at java.base/java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:188)
  
  	at org.apache.pulsar.client.admin.PulsarAdminException.wrap(PulsarAdminException.java:252)
  	at org.apache.pulsar.client.admin.internal.BaseResource.sync(BaseResource.java:352)
  	at org.apache.pulsar.client.admin.internal.LookupImpl.lookupPartitionedTopic(LookupImpl.java:67)
  	at org.apache.pulsar.broker.loadbalance.extensions.ExtensibleLoadManagerCloseTest.testCloseAfterLoadingBundles(ExtensibleLoadManagerCloseTest.java:105)
  	at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:103)
  	at java.base/java.lang.reflect.Method.invoke(Method.java:580)
  	at org.testng.internal.invokers.MethodInvocationHelper.invokeMethod(MethodInvocationHelper.java:139)
  	at org.testng.internal.invokers.InvokeMethodRunnable.runOne(InvokeMethodRunnable.java:47)
  	at org.testng.internal.invokers.InvokeMethodRunnable.call(InvokeMethodRunnable.java:76)
  	at org.testng.internal.invokers.InvokeMethodRunnable.call(InvokeMethodRunnable.java:11)
  	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:317)
  	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1144)
  	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:642)
  	at java.base/java.lang.Thread.run(Thread.java:1583)
```

### Are you willing to submit a PR?

- [ ] I'm willing to submit a PR!

Recent:

### Example failures

- [2025-09-18T07:22:04.7670041Z](https://github.com/apache/pulsar/actions/runs/17817810114/job/50661792602#step:8:2318) 


### Exception stacktrace

```
org.apache.pulsar.client.admin.PulsarAdminException: java.lang.IllegalStateException: Client instance has been closed.
	at org.apache.pulsar.client.admin.PulsarAdminException.wrap(PulsarAdminException.java:252)
	at org.apache.pulsar.client.admin.internal.BaseResource.sync(BaseResource.java:354)
	at org.apache.pulsar.client.admin.internal.TopicsImpl.createPartitionedTopic(TopicsImpl.java:306)
	at org.apache.pulsar.client.admin.Topics.createPartitionedTopic(Topics.java:475)
	at org.apache.pulsar.broker.loadbalance.extensions.ExtensibleLoadManagerCloseTest.testCloseAfterLoadingBundles(ExtensibleLoadManagerCloseTest.java:90)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:569)
	at org.testng.internal.invokers.MethodInvocationHelper.invokeMethod(MethodInvocationHelper.java:139)
	at org.testng.internal.invokers.InvokeMethodRunnable.runOne(InvokeMethodRunnable.java:47)
	at org.testng.internal.invokers.InvokeMethodRunnable.call(InvokeMethodRunnable.java:76)
	at org.testng.internal.invokers.InvokeMethodRunnable.call(InvokeMethodRunnable.java:11)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
	at java.base/java.lang.Thread.run(Thread.java:840)
	Suppressed: org.apache.pulsar.client.admin.PulsarAdminException: java.lang.IllegalStateException: Client instance has been closed.
		at org.apache.pulsar.client.admin.internal.BaseResource.getApiException(BaseResource.java:300)
		... 16 more
	Caused by: java.lang.IllegalStateException: Client instance has been closed.
		at org.glassfish.jersey.internal.guava.Preconditions.checkState(Preconditions.java:169)
		at org.glassfish.jersey.client.JerseyClient.checkNotClosed(JerseyClient.java:248)
		at org.glassfish.jersey.client.JerseyWebTarget.checkNotClosed(JerseyWebTarget.java:118)
		at org.glassfish.jersey.client.JerseyWebTarget.path(JerseyWebTarget.java:129)
		at org.glassfish.jersey.client.JerseyWebTarget.path(JerseyWebTarget.java:38)
		at org.apache.pulsar.client.admin.internal.TopicsImpl.topicPath(TopicsImpl.java:1224)
		at org.apache.pulsar.client.admin.internal.TopicsImpl.createPartitionedTopicAsync(TopicsImpl.java:337)
		at org.apache.pulsar.client.admin.internal.TopicsImpl.createPartitionedTopicAsync(TopicsImpl.java:330)
```

--- a/pulsar-broker/src/main/java/org/apache/pulsar/broker/loadbalance/extensions/store/TableViewLoadDataStoreImpl.java
+++ b/pulsar-broker/src/main/java/org/apache/pulsar/broker/loadbalance/extensions/store/TableViewLoadDataStoreImpl.java
@@ -18,6 +18,7 @@
  */
 package org.apache.pulsar.broker.loadbalance.extensions.store;
 
+import com.google.common.annotations.VisibleForTesting;
 import java.io.IOException;
 import java.util.Map;
 import java.util.Optional;
@@ -145,7 +146,14 @@ public class TableViewLoadDataStoreImpl<T> implements LoadDataStore<T> {
     public synchronized void closeTableView() throws IOException {
         validateState();
         if (tableView != null) {
-            tableView.close();
+            // Close asynchronously without waiting: the close future may only be able to complete on the
+            // client's internal executor thread, and that thread can be blocked on this store's monitor
+            // (e.g. a ServiceUnitStateChannel StateChangeListener calling pushAsync/removeAsync), so a
+            // blocking close while holding the monitor could deadlock.
+            tableView.closeAsync().exceptionally(e -> {
+                log.warn().attr("topic", topic).exception(e).log("Failed to close table view");
+                return null;
+            });
             tableView = null;
         }
     }
@@ -160,7 +168,11 @@ public class TableViewLoadDataStoreImpl<T> implements LoadDataStore<T> {
     private synchronized void closeProducer() throws IOException {
         validateState();
         if (producer != null) {
-            producer.close();
+            // Close asynchronously without waiting; see closeTableView() for the deadlock rationale.
+            producer.closeAsync().exceptionally(e -> {
+                log.warn().attr("topic", topic).exception(e).log("Failed to close producer");
+                return null;
+            });
             producer = null;
         }
     }
@@ -209,6 +221,11 @@ public class TableViewLoadDataStoreImpl<T> implements LoadDataStore<T> {
     public synchronized void shutdown() throws IOException {
         close();
         isShutdown = true;
+    }
+
+    @VisibleForTesting
+    synchronized void setTableView(TableView<T> tableView) {
+        this.tableView = tableView;
     }
 
     private String validateProducer() {

--- a/pulsar-broker/src/test/java/org/apache/pulsar/broker/loadbalance/extensions/store/LoadDataStoreTest.java
+++ b/pulsar-broker/src/test/java/org/apache/pulsar/broker/loadbalance/extensions/store/LoadDataStoreTest.java
@@ -18,9 +18,12 @@
  */
 package org.apache.pulsar.broker.loadbalance.extensions.store;
 
+import static org.mockito.Mockito.doAnswer;
+import static org.mockito.Mockito.mock;
 import static org.mockito.Mockito.spy;
 import static org.mockito.Mockito.times;
 import static org.mockito.Mockito.verify;
+import static org.mockito.Mockito.when;
 import static org.testng.Assert.assertEquals;
 import static org.testng.Assert.assertFalse;
 import static org.testng.Assert.assertThrows;
@@ -29,12 +32,15 @@ import com.google.common.collect.Sets;
 import java.util.HashMap;
 import java.util.Map;
 import java.util.UUID;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.CountDownLatch;
 import lombok.AllArgsConstructor;
 import lombok.Cleanup;
 import lombok.Data;
 import lombok.NoArgsConstructor;
 import org.apache.commons.lang3.reflect.FieldUtils;
 import org.apache.pulsar.broker.auth.MockedPulsarServiceBaseTest;
+import org.apache.pulsar.client.api.TableView;
 import org.apache.pulsar.common.naming.NamespaceName;
 import org.apache.pulsar.common.naming.TopicDomain;
 import org.apache.pulsar.common.policies.data.TenantInfoImpl;
@@ -186,6 +192,53 @@ public class LoadDataStoreTest extends MockedPulsarServiceBaseTest {
 
         loadDataStore.pushAsync("2", 2).get();
         loadDataStore.removeAsync("2").get();
+    }
+
+    @Test(timeOut = 30_000)
+    @SuppressWarnings("unchecked")
+    public void testShutdownDoesNotDeadlockWithConcurrentStoreAccess() throws Exception {
+        String topic = TopicDomain.persistent + "://" + NamespaceName.SYSTEM_NAMESPACE + "/" + UUID.randomUUID();
+        var loadDataStore =
+                (TableViewLoadDataStoreImpl<Integer>) LoadDataStoreFactory.create(pulsar, topic, Integer.class);
+        loadDataStore.start();
+        loadDataStore.closeTableView();
+
+        // Replace the table view with a stub whose close future completes only after another thread has
+        // called a synchronized method of the store. This mirrors the production interleaving where the
+        // reader close future completes on the client's internal executor thread while a channel
+        // StateChangeListener on that same thread is blocked in removeAsync() waiting for the store
+        // monitor held by the closing thread.
+        CompletableFuture<Void> closeSignal = new CompletableFuture<>();
+        CountDownLatch closeStarted = new CountDownLatch(1);
+        TableView<Integer> stubTableView = mock(TableView.class);
+        when(stubTableView.closeAsync()).thenAnswer(invocation -> {
+            closeStarted.countDown();
+            return closeSignal;
+        });
+        doAnswer(invocation -> {
+            closeStarted.countDown();
+            closeSignal.get();
+            return null;
+        }).when(stubTableView).close();
+        loadDataStore.setTableView(stubTableView);
+
+        Thread concurrentAccess = new Thread(() -> {
+            try {
+                closeStarted.await();
+                loadDataStore.removeAsync("key");
+                closeSignal.complete(null);
+            } catch (InterruptedException e) {
+                Thread.currentThread().interrupt();
+            }
+        });
+        concurrentAccess.start();
+        try {
+            loadDataStore.shutdown();
+        } finally {
+            concurrentAccess.join(5000);
+        }
+        assertFalse(concurrentAccess.isAlive());
+        assertTrue(closeSignal.isDone());
     }
 
     @Test

